File: javascript/code-challenges/stacksAndQueues/classes/stack.js
'use strict';

const Node = require('./node.js');

class Stack {
  constructor() {
    this.top = null;
    this.storage = new Array();
  }


  push(item) {
    let node = new Node(item);
    this.storage.push(node.value);
    this.top = node.value;
  }

  pop() {
    if (this.storage.length < 1) {
      this.top = null;
      
      return 'Stack Empty!';
    } else {
      let node = this.storage.pop();
      console.log(this.storage);
      return node;
    }
  }

  peek() {
    if (this.storage.length < 1) {
      return 'Stack Empty!';
    } else {
      return this.top;
    }
  }

  isEmpty(){
    let empty = true;
    if (this.storage.length < 1) {
      return empty;
    } else {
      empty = false;
      return empty;
    }
  }
}

module.exports = Stack;

File: python/code_challenges/stack_queue_animal_shelter/stack.py
from node import Node

class Stack:
    '''
    The stack data structure, which stores nodes in a First In Last Out format.
    Attributes:
        self.top: Last node in stack
    '''
    def __init__(self, top = None):
        self.top = top

    def __str__(self):
        return f'A Stack, its top holds {self.top}'

    def __repr__(self):
        return f'Stack(top: Node({self.front}))'

    def push(self, value):
        '''
        Adds a new node with input value to the top of the stack with an O(1) Time performance.
        input: value as any value
        output: None
        '''
        node = Node(value)

        if self.top == None:
            self.top = node
        else:
            temp_top = node
            temp_top.next = self.top
            self.top = temp_top

    def pop(self):
        '''
        Returns the value of the node from the top of the stack. Removes the node from the top of the stack. Should raise exception when called on empty stack
        input: None
        output: value of Node from top of stack
        '''
        if self.top == None:
            raise Exception('Stack is empty.')
        ret_node = self.top
        self.top = self.top.next

        return ret_node.value

    def peek(self):
        '''
        Returns the value of the node located at the top of the stack. Should raise exception when called on empty stack.
        input: None
        output: Value of top node of stack
        '''

        if self.top == None:
            raise Exception('Stack is empty.')
        else:
            return self.top.value

    def is_empty(self):
        '''
        Returns a boolean indicating whether or not the stack is empty
        input: None
        output: Boolean
        '''
        if self.top:
            return False
        else:
            return True
